### PR TITLE
fix(rust): fix build for wasm

### PR DESCRIPTION
**Pull request by lorepozo**
Fixes issue #8783, where v0.29 broke the build for wasm. In addition to fixing the build, this adds a CI test to prevent regressions like this in the future. I've gone ahead and added every feature that works with wasm (not all features work out of the box and adding support for those can be done in the future as needed).

Relatedly there's an issue for the `home` crate rust-lang/cargo#12297 to make it compile for wasm.

### CODE REVIEW

**Comment by lorepozo** [comment]
@ritchie46 would you review this change? Build failures are unrelated (probably just need to rebase)

File: polars/polars-io/src/utils.rs [hunk 1]
```diff
@@ -16,7 +15,9 @@ use crate::ArrowSchema;
 pub fn resolve_homedir(path: &Path) -> PathBuf {
     // replace "~" with home directory
     if path.starts_with("~") {
-        if let Some(homedir) = home_dir() {
+        // home crate does not compile on wasm https://github.com/rust-lang/cargo/issues/12297
+        #[cfg(not(target_arch = "wasm32"))]
+        if let Some(homedir) = home::home_dir() {
             return homedir.join(path.strip_prefix("~").unwrap());
         }
     }
```
Comment (universalmind303) [inline polars/polars-io/src/utils.rs:23]: this is probably fine if we are just concerned with _compiling_, but since WASM doesn't have a filesystem, i'm wondering if `resolve_homedir` should be feature flagged instead? Likely anything that calls `resolve_homedir` will panic on `wasm32` anyways.

File: .github/workflows/test-rust.yml [hunk 2]
```diff
@@ -112,3 +112,117 @@ jobs:
...
+            -F asof_join
+            -F avro
+            -F avx512
+            -F bigidx
+            -F binary_encoding
+            -F checked_arithmetic
+            -F chunked_ids
+            -F coalesce
+            -F concat_str
+            -F concat_str
```
Comment (highway900) [inline .github/workflows/test-rust.yml:139]: duplicate feature

Comment (lorepozo) [reply .github/workflows/test-rust.yml]: oops, thanks! fixed.

**Comment by lorepozo** [comment]
@ritchie46 — would you please review this? I'm stuck on 0.28 because the build has been broken for wasm.

File: .github/workflows/test-rust.yml [hunk 3]
```diff
@@ -112,3 +112,116 @@ jobs:
...
+  build-other-targets:
+    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: true
+      matrix:
+        include:
+        - name: Wasm
+          target: wasm32-unknown-unknown
+          features: >
+            -F abs
```
Comment (ritchie46) [inline .github/workflows/test-rust.yml:125]: Not that familiar with this. What does -F mean in this context?

File: polars/polars-utils/src/wasm.rs [hunk 4]
```diff
@@ -5,6 +5,14 @@ impl Pool {
         rayon::current_num_threads()
     }
 
+    pub fn current_thread_index(&self) -> Option<usize> {
```
Comment (ritchie46) [inline polars/polars-utils/src/wasm.rs:8]: Why was this added?

File: .github/workflows/test-rust.yml (hunk 3 above)
Comment (lorepozo) [reply .github/workflows/test-rust.yml:125]: this is used further down on line 226 — it specifies a feature:

```sh
cargo build --target ${{ matrix.target }} -p polars --no-default-features ${{ matrix.features }}
```

File: polars/polars-utils/src/wasm.rs (hunk 4 above)
Comment (lorepozo) [reply polars/polars-utils/src/wasm.rs:8]: it's used in [`polars-core/src/frame/groupby/hashing.rs`](https://github.com/pola-rs/polars/blob/main/polars/polars-core/src/frame/groupby/hashing.rs#L29) and [`polars-core/src/series/implementations/struct_.rs`](https://github.com/pola-rs/polars/blob/main/polars/polars-core/src/series/implementations/struct_.rs#L283).

For added context, [`crate::POOL`](https://github.com/pola-rs/polars/blob/main/polars/polars-core/src/lib.rs#L65) is configured to a custom struct when building on wasm rather than the corresponding lazily-built `rayon::ThreadPool`.

**Comment by lorepozo** [comment]
Just rebased, no manual changes were necessary. @ritchie46 please take another look! Would like to fix the regression from 0.28.

**Comment by lorepozo** [comment]
rebased to deal with conflict from #10279 in `crates/polars-io/Cargo.toml`

**Comment by lorepozo** [comment]
Rebased again — @orlp perhaps you can review?

**Review by universalmind303 — APPROVED** [review]
seems reasonable to me.

only suggestion I have is maybe moving the ci stuff to the Makefile so it is easier to build outside of CI as well.

**Comment by lorepozo** [comment]
Thanks! It appears the Makefile is only used for working in py-polars, but the relevant libraries are part of the rust crates. I'm happy to put the ci stuff in a more appropriate place — I think the next best place to list out the features would be in `Cargo.toml` as a new feature like `wasm-all`, but that would add to the public API of the crate for the purposes of testing, which is usually a pattern to avoid. So I'm leaning towards keeping it in CI. I'm open to other ideas if there's another good place! (Or if you think a feature like `wasm-all` would be preferable to keeping the list in CI, I can move it there)

**Comment by orlp** [comment]
@lorepozo 

> Thanks! It appears the Makefile is only used for working in py-polars, but the relevant libraries are part of the rust crates.

There is another Makefile in the `crates` folder, it can go there.

**Comment by lorepozo** [comment]
There are a few other ways that come to mind:
1. support all features on wasm
2. have a smaller explicit list of supported features on wasm.
3. use a regression testing framework that only fails version N+1 if version N worked under the same conditions (target + features).

I haven't looked into the difficulty of the first choice. The second choice risks regressions when other features lose support, and is hence tied to a judgement call I'd rather not make. I don't know of solutions for the third choice, though it seems technically solvable with the limited scope of cargo features across targets.

And just pushed the change for the Makefile.

**Comment by lorepozo** [comment]
Actually, another option is to have a list of _excludes_ — features that are _not_ supported on wasm. That list would be smaller and more precisely capture where the problems exist. I'll try making that change now.

**Comment by lorepozo** [comment]
Nice, it looks much better.

In case you're curious, most of the build failures are coming from `mio`.

**Comment by lorepozo** [comment]
I can submit a small PR after this which will fix many of the `mio` issues so that `dtype-full` works on wasm. The problem stems from `polars-pipe` unconditionally depends on `ipc`/`async` of `polars-io`.

**Comment by orlp** [comment]
@lorepozo I am reverting this because it broke our CI.

You can view the error log here: https://github.com/pola-rs/polars/actions/runs/5880942481/job/15948348046?pr=10388.

Could you make a new pull request, and test it first in CI on a fork?

**Comment by lorepozo** [comment]
It was tested correctly in CI via this PR, looks like a merge race because of #10492 breaking the `decompress` feature on wasm. I'll create a new PR with that feature excluded as well.
